services/streammanager/kafka/kafkamanager.go

```diff
@@ -31,320 +33,385 @@ type Config struct {
 	if err != nil {
-		return nil, fmt.Errorf("[Kafka] Error while unmarshalling dest config :: %w", err)
-	}
-	hosts := make([]string, 0)
-	hostNames := strings.Split(destConfig.HostName, ",")
-	for _, hostName := range hostNames {
-		hosts = append(hosts, hostName+":"+destConfig.Port)
+		return nil, fmt.Errorf("[Kafka] Error while unmarshalling destination configuration %+v, got error: %w",
+			destConfigJSON, err,
+		)
 	}
-	isSslEnabled := destConfig.SslEnabled
 
-	conf := getDefaultConfiguration()
-	conf.Producer.Timeout = o.Timeout
+	if err = destConfig.validate(); err != nil {
+		return nil, fmt.Errorf("invalid configuration: %w", err)
+	}
 
-	if isSslEnabled {
-		caCertificate := destConfig.CACertificate
-		conf.Net.TLS.Enable = true
-		if caCertificate != "" {
-			tlsConfig := newTLSConfig(caCertificate)
-			if tlsConfig != nil {
-				conf.Net.TLS.Config = tlsConfig
+	clientConf := client.Config{
+		DialTimeout: kafkaDialTimeout,
+	}
+	if destConfig.SslEnabled {
+		if destConfig.CACertificate != "" {
+			clientConf.TLS = &client.TLS{
+				CACertificate: []byte(destConfig.CACertificate),
+				Cert:          clientCert,
+				Key:           clientKey,
 			}
+		} else {
+			clientConf.TLS = &client.TLS{WithSystemCertPool: true}
 		}
-		if destConfig.UseSASL {
-			// SASL is enabled only with SSL
-			err = setSASLConfig(conf, destConfig)
+
+		if destConfig.UseSASL { // SASL is enabled only with SSL
+			clientConf.SASL = &client.SASL{
+				Username: destConfig.Username,
+				Password: destConfig.Password,
+			}
+			clientConf.SASL.ScramHashGen, err = client.ScramHashGeneratorFromString(destConfig.SaslType)
 			if err != nil {
-				return nil, fmt.Errorf("[Kafka] Error while setting SASL config :: %w", err)
+				return nil, fmt.Errorf("[Kafka] invalid SASL type: %w", err)
 			}
 		}
 	}
 
-	return sarama.NewSyncProducer(hosts, conf)
-}
-
-// Sets SASL authentication config for Kafka
-func setSASLConfig(config *sarama.Config, destConfig Config) (err error) {
-	config.Net.SASL.Enable = true
-	config.Net.SASL.User = destConfig.Username
-	config.Net.SASL.Password = destConfig.Password
-	config.Net.SASL.Handshake = true
-	switch destConfig.SaslType {
-	case "plain":
-		config.Net.SASL.Mechanism = sarama.SASLTypePlaintext
-	case "sha512":
-		config.Net.SASL.SCRAMClientGeneratorFunc = func() sarama.SCRAMClient {
-			return &xdgSCRAMClient{HashGeneratorFcn: SHA512}
+	var c *client.Client
+	// The only component that supports multiple hosts is the consumer which we are not using here.
+	// It is also worth mentioning that internally the list of brokers should be populated accordingly anyway given
+	// that Kafka uses the "advertised.listeners" property to let clients know where they can connect.
+	// see: https://kafka.apache.org/documentation/#brokerconfigs_advertised.listeners
+	for _, host := range strings.Split(destConfig.HostName, ",") {
+		c, err = client.New("tcp", host+":"+destConfig.Port, clientConf)
```

Comment:
Here we are only connecting to the first available host and attempting to send events to the host. If the host goes down all the events would fail since we call NewProducer only once for every destination type right. Is this part of the code consistent with the sarama implementation?

Reply:
@cisse21 have you read the comment right above the line you're commenting on? 

Consider that these servers are just used for the initial connection to discover the full cluster membership (which may change dynamically) so what we pass there does not need to contain the full set of servers as it will be discovered anyway. The problem would be if we can't connect  and don't try the others but that's not what we're doing here, right?

Reply:
I read the comment but just wanted to make sure what I understood was right. Thanks for the detailed reply 👍

Reply:
@cisse21 can you have another looksie please? I made a small change to reintroduce the brokers list [here](https://github.com/rudderlabs/rudder-server/pull/1920/commits/f438dd0a7124df1af59a2794bf8582b528311624).